File: go/2015/day06a/day06a.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

const (
	MAX_ROWS = 1000
	MAX_COLS = 1000
)

func usage(progname string) {
	fmt.Fprintf(os.Stderr, "usage: %s <input file>\n", progname)
	os.Exit(1)
}

func perform(grid *[MAX_ROWS][MAX_COLS]bool, action string, r1 int, c1 int, r2 int, c2 int) {
	for row := r1; row <= r2; row++ {
		for col := c1; col <= c2; col++ {
			switch action {
			case "turn on":
				grid[row][col] = true
			case "turn off":
				grid[row][col] = false
			case "toggle":
				grid[row][col] = !grid[row][col]
			}
		}
	}
}

func count(grid [MAX_ROWS][MAX_COLS]bool) uint32 {
	count := uint32(0)
	for row := 0; row < MAX_ROWS; row++ {
		for col := 0; col < MAX_COLS; col++ {
			if grid[row][col] {
				count++
			}
		}
	}
	return count
}

func process(filename string) uint32 {
	var grid [MAX_ROWS][MAX_COLS]bool
	re := regexp.MustCompile(`(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)`)

	file, err := os.Open(filename)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		matches := re.FindStringSubmatch(line)
		if matches != nil {
			action := matches[1]
			r1, _ := strconv.Atoi(matches[2])
			c1, _ := strconv.Atoi(matches[3])
			r2, _ := strconv.Atoi(matches[4])
			c2, _ := strconv.Atoi(matches[5])
			perform(&grid, action, r1, c1, r2, c2)
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintf(os.Stderr, "Error reading file: %v\n", err)
		os.Exit(1)
	}

	return count(grid)
}

func main() {
	if len(os.Args) < 2 {
		usage(os.Args[0])
	}

	filename := os.Args[1]
	result := process(filename)
	fmt.Printf("result = %d\n", result)
}

File: dart/2015/day06a/day06a.dart
#!/usr/bin/env dart

import 'dart:io';

const int MAX_ROW = 1000;
const int MAX_COL = 1000;

void usage(String progName) {
    print('usage: $progName <input file>');
    exit(1);
}

void perform(List<bool> grid, String action, int r1, int c1, int r2, int c2) {
    for ( var row = r1; row <= r2; row++ ) {
        for ( var col = c1; col <= c2; col++ ) {
            var index = row * MAX_COL + col;
            switch ( action ) {
            case 'turn on':
                grid[index] = true;
                break;
            case 'turn off':
                grid[index] = false;
                break;
            case 'toggle':
                grid[index] = !grid[index];
                break;
            }
        }
    }
}

int count(List<bool> grid) {
    var total = 0;
    for ( var row = 0; row < MAX_ROW; row++ ) {
        for ( var col = 0; col < MAX_COL; col++ ) {
            var index = row * MAX_COL + col;
            if ( grid[index] ) {
                total++;
            }
        }
    }
    return total;
}

int process(String filename) {
    var grid = List<bool>.filled(MAX_ROW * MAX_COL, false);
    final lines = File(filename).readAsLinesSync();
    for ( final line in lines ) {
        var re = RegExp(r'(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)');
        var match = re.firstMatch(line);
        if ( match != null ) {
            var action = match.group(1)!;
            var r1 = int.parse(match.group(2)!);
            var c1 = int.parse(match.group(3)!);
            var r2 = int.parse(match.group(4)!);
            var c2 = int.parse(match.group(5)!);
            perform(grid, action, r1, c1, r2, c2);
        }
    }
    return count(grid);
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final filename = args[0];
    final result = process(filename);
    print('result = $result');
}
